### PR TITLE
Time-invariant event log

### PR DESCRIPTION
**Pull request by pierluca**
This pull requests deals with #2376 and adds a few documentation fixes.
It makes the event log contract time-invariant by referring exclusively to the current block timestamp rather than `time.Now()`

I'm still looking into the unit-tests.

🙅‍ Friendly checklist:

- [x] 0. Code comments are added (or updated) when/where needed and explain the WHY of the code.
- [x] 1. Design choices, user documentation and any additional doc are added (or updated) in READMEs.
- [ ] 2. Any new behaviour is tested and small units of code that can be are unit tested.
- [ ] 3. Code comments are added on tests to explain what they do.
- [ ] 4. Errors are systematically wrapped with a meaningful message using `xerrors.Errorf` and the `%v` verb.
- [ ] 5. Hard limit of 80 chars is always respected.
- [x] 6. Changes are backward compatible.
- [x] 7. Indentation level does not exceed 5, although 4 is already suspicious.
- [x] 8. Functions, files, and packages are kept to a manageable size and decomposed into smaller units if needed.
- [x] 9. There are no magic values.

### CODE REVIEW

**Comment by ineiti** [comment]
Draft-comments: I started doing 2 separate PRs for separating:
- comments changes that don't change the code
- actual implementation of what I'm doing
Of course I do this only once I'm done with the PR itself...

**Comment by pierluca** [comment]
I'm waiting for Travis to work, based on : https://github.com/dedis/cothority/pull/2383 and https://github.com/dedis/cothority/issues/2367

File: eventlog/service.go [hunk 1]
```diff
@@ -143,12 +148,11 @@ filter:
...
 }
 
-func decodeAndCheckEvent(coll byzcoin.ReadOnlyStateTrie, eventBuf []byte) (*Event, error) {
-	// Check the timestamp of the event: it should never be in the future,
-	// and it should not be more than 30 seconds in the past. (Why 30 sec
-	// and not something more auto-scaling like blockInterval * 30?
-	// Because a # of blocks limit is too fragile when using fast blocks for
-	// tests.)
+func decodeAndCheckEvent(rst byzcoin.ReadOnlyStateTrie, eventBuf []byte) (*Event, error) {
+	// Check the timestamp of the event: it should never be in the future beyond the current block,
```
Comment (ineiti) [inline eventlog/service.go:152]: you accept `4 * config.BlockInterval`, which is fine, but doesn't match your comment here.

File: eventlog/service.go [hunk 2]
```diff
@@ -20,11 +20,16 @@ const contractName = "eventlog"
...
-// this will pretect the buckets from getting too big. During low message
+// this will protect the buckets from getting too big. During low message
 // arrival (< 1 per 5 sec) it does not create extra buckets, because time
 // periods with no events do not need buckets created for them.
 const bucketMaxAge = 5 * time.Second
 
+// eventDelayTolerance indicates how much time after the event timestamp
+// we are stilling willing to accept an event as genuine and contemporaneous,
+// and hence as valid. Expressed in [ns]
+const eventDelayTolerance = time.Duration(-60 * 1e9)
```
Comment (ineiti) [inline eventlog/service.go:31]:
```suggestion
const eventDelayTolerance = time.Minute
```

And reverse the comment to say _before_, as well as changing the code.

**Review by ineiti — CHANGES_REQUESTED** [review]
Are there tests for this change?

**Comment by ineiti** [comment]
You can rebase to master, the updates to the testing framework should make it work more reliable now. And if new tests fail, it'll be easier to make them pass...

**Comment by ineiti** [comment]
In fact the latest travis test failed because of an error in coveralls...

**Comment by ineiti** [comment]
@pierluca bumping this, as the `byzcoin.BCTest` is now in `master` and produces quite stable tests.

**Comment by pierluca** [comment]
Thanks for the information @ineiti.
I'm committed to seeing this through, but it's not my #1 priority right now.
I'll come back to this as soon as I can find some spare time.

**Comment by ineiti** [comment]
Is this something you think working on, or should it be closed? You might want to delegate it ;)

**Comment by pierluca** [comment]
Considering Cothority is reaching C4DT support's end-of-life in April 2022 and that this feature is not currently used by any production system, I'm closing this PR. I'd be happy to pick it up again later should it become relevant / a priority.

**Pull request closed without merging**